okex-price/src/pricefeed/feeder.rs

```diff
@@ -0,0 +1,107 @@
+use super::{error::PriceFeedError, PriceFeedConfig};
+use futures::{
+    stream::{SplitSink, SplitStream},
+    SinkExt, Stream, StreamExt,
+};
+use serde::{Deserialize, Serialize};
+use shared::{
+    payload,
+    pubsub::{Envelope, MessagePayload},
+};
+use std::pin::Pin;
+use tokio::net::TcpStream;
+use tokio_tungstenite::{connect_async, tungstenite::Message, MaybeTlsStream, WebSocketStream};
+use url::Url;
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct ChannelData {
+    channel: String,
+    inst_id: String,
+}
+
+#[derive(serde::Serialize)]
+struct ChannelSubscription {
+    op: String,
+    args: Vec<ChannelData>,
+}
+
+impl ChannelSubscription {
+    fn new() -> Self {
+        Self {
+            op: "subscribe".to_string(),
+            args: vec![ChannelData {
+                channel: "tickers".to_string(),
+                inst_id: "BTC-USD-SWAP".to_string(),
+            }],
+        }
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct TickersChannelData {
+    inst_type: String,
+    inst_id: String,
+    last: String,
+    last_sz: String,
+    ask_px: String,
+    ask_sz: String,
+    bid_px: String,
+    bid_sz: String,
+    open24h: String,
+    high24h: String,
+    low24h: String,
+    vol_ccy24h: String,
+    vol24h: String,
+    sod_utc0: String,
+    sod_utc8: String,
+    ts: String,
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub struct OkexPriceTick {
+    arg: ChannelData,
+    data: Vec<TickersChannelData>,
+}
+
+payload! { OkexPriceTick, "okex.price" }
+
+#[derive(Debug)]
+pub struct PriceFeed {
+    sender: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+    receiver: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+}
+
+impl PriceFeed {
+    pub async fn new() -> Result<Self, PriceFeedError> {
+        let config = PriceFeedConfig {
+            url: Url::parse("wss://ws.okx.com:8443/ws/v5/public").unwrap(),
+        };
+        let (ws_stream, _) = connect_async(config.url).await?;
+        let (sender, receiver) = ws_stream.split();
+
+        Ok(Self { sender, receiver })
+    }
+
+    pub async fn subscribe_btcusdswap(
+        &mut self,
+    ) -> Result<Pin<Box<dyn Stream<Item = Envelope<OkexPriceTick>> + Send>>, PriceFeedError> {
```

Comment:
`Envelope` is for `pubsub` serialisation - not what re receive from okex.

Reply:
I thought it would be a good idea to have metadata and payload type on the data received from okex but I will remove it and deserialize `OkexPriceTick` without the wrapper.